cmd/tailscale/main.go

```diff
@@ -458,8 +462,22 @@ func (a *App) runBackend(ctx context.Context) error {
 			case ExitAllowLANEvent:
 				state.Prefs.ExitNodeAllowLANAccess = bool(e)
 				go b.backend.SetPrefs(state.Prefs)
-			case WebAuthEvent:
-				if !signingIn {
+			case WebAuthAndMaybeSetServerEvent:
+				if e.URL != "" && !signingIn {
+					state.Prefs.ControlURL = e.URL
+					b.backend.SetPrefs(state.Prefs)
```

Comment:
This is disk I/O on the UI thread, no? Note line 460 does this in a goroutine.

Should this move to line 474?

But you'll want to make a copy of state.Prefs first to close over in that goroutine, otherwise it'd be a data race reading state.Prefs from another goroutine.  (on line 460 it's not a problem because the immediate "go F(arg)" copies arg to the arguments of F before the goroutine starts)